components/org.wso2.carbon.identity.application.authenticator.adapter/src/main/java/org/wso2/carbon/identity/application/authenticator/adapter/AbstractAuthenticatorAdapter.java

```diff
@@ -0,0 +1,181 @@
+import org.wso2.carbon.identity.application.authentication.framework.AuthenticatorFlowStatus;
+import org.wso2.carbon.identity.application.authentication.framework.context.AuthenticationContext;
+import org.wso2.carbon.identity.application.authentication.framework.exception.AuthenticationFailedException;
+import org.wso2.carbon.identity.application.authentication.framework.exception.LogoutFailedException;
+import org.wso2.carbon.identity.application.authentication.framework.util.FrameworkConstants;
+import org.wso2.carbon.identity.application.authenticator.adapter.internal.AuthenticatorAdapterDataHolder;
+import org.wso2.carbon.identity.application.authenticator.adapter.util.AuthenticatorAdapterConstants;
+import org.wso2.carbon.identity.application.common.model.Property;
+import org.wso2.carbon.identity.base.AuthenticatorPropertyConstants;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * This class holds the external custom authentication.
+ */
+public abstract class AbstractAuthenticatorAdapter extends AbstractApplicationAuthenticator {
+
+    private static final Log log = LogFactory.getLog(AbstractAuthenticatorAdapter.class);
+    protected String authenticatorName;
+    protected String friendlyName;
+
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+
+        return true;
+    }
+
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request, HttpServletResponse response,
+                                           AuthenticationContext context)
+            throws AuthenticationFailedException, LogoutFailedException {
+
+        if (context.isLogoutRequest()) {
+            return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+        }
+
+        context.removeProperty(AuthenticatorAdapterConstants.EXECUTION_STATUS_PROP_NAME);
+        Map<String, Object> eventContext = new HashMap<>();
+        eventContext.put(AuthenticatorAdapterConstants.AUTH_REQUEST, request);
+        eventContext.put(AuthenticatorAdapterConstants.AUTH_RESPONSE, response);
+        eventContext.put(AuthenticatorAdapterConstants.AUTH_CONTEXT, context);
+        ActionExecutionStatus executionStatus = executeAction(context, eventContext, context.getTenantDomain());
+        context.setProperty(AuthenticatorAdapterConstants.EXECUTION_STATUS_PROP_NAME, executionStatus);
+
+        if (executionStatus.getStatus() == ActionExecutionStatus.Status.INCOMPLETE) {
+            if (authenticatorName.equals(context.getProperty(FrameworkConstants.LAST_FAILED_AUTHENTICATOR))) {
+                context.setRetrying(true);
+            }
+            context.setCurrentAuthenticator(getName());
+            context.setRetrying(false);
+            return AuthenticatorFlowStatus.INCOMPLETE;
```

Comment:
I don't think this is needed.
When it comes to custom authenticator it's them who implement the authentication. So whether retrying or not is their choice and they will have to interact with the user accordingly